### ISSUE TITLE
Add publish method to RxClient that returns Single<MqttPublish>

### ISSUE THREAD

**Issue by tchupp** [hivemq/hivemq-mqtt-client#262]
## Problem or use case

Tracking latency metrics for individual MQTT publish requests.

Right now, it is difficult to track these metrics for a specific publish request, due to the publish method on the RxClient accepting a `Flowable<MqttPublish>`.

Currently, we are working around this by using the AsyncClient in certain spots, but we would prefer to be able to use the RxClient everywhere.

Here is pseudo code for how we are working around:

```java
...
Timer timer = Timer.start();

return Single.fromFuture(mqttAsyncClient.publish(message))
  .doOnSuccess(message -> timer.stopWithSuccess())
  .doOnError(error -> timer.stopWithFailure())
...
```

## Preferred solution or suggestions

It would be nice to be able to do something like this:

```java
...
Timer timer = Timer.start();

return mqttRxClient.publish(message)
  .doOnSuccess(message -> timer.stopWithSuccess())
  .doOnError(error -> timer.stopWithFailure())
...
```

Using a `Single<T>` as the return type might also be a bit more ergonomic than the `MqttPublishResult`, as the `Single` provides nice methods for error handling, where the `MqttPublishResult` has any potential errors available as an `Optional<Exception>`

**Comment by SgtSilvio** [comment]
First of all sorry for the delayed response.

The use of Flowable in the Rx API is intended, as this is how Rx works. There are outgoing and incoming asynchronous streams of messages.
Publishing each message separately is possible but not entirely reactive, as you can not take advantage of the non blocking reactive back pressure mechanism.
Publishing each message separately in a very tight loop may result in blocking the calling thread. (This is probably not your use case)

We do not recommend to use your workaround as `Single.fromFuture` blocks on the subscribing thread.
Instead transforming the Flowable to a Single can easily be done with the Rx API:
```java
rxClient.publish(Flowable.just(message))
        .singleOrError()
```
Using the error callback is then also simple:
```java
rxClient.publish(Flowable.just(message))
        .singleOrError()
        .flatMap(publishResult -> publishResult.getError()
                .map(Single::error)
                .orElse(Single.just(publishResult)));
```

Another thing to mention: The use of the timer above is not correct, as a Single (and all other reactive types) are lazy. This means you should not start the timer and then return a reactive type, as the publish operation is not started immediately but only when subscribing to the Single. You even can subscribe multiple times to the same returned Single. Then the metrics are completely wrong.
So you should to do something like:
```java
rxClient.publish(Flowable.just(message))
        .singleOrError()
        .flatMap(publishResult -> publishResult.getError()
                .map(Single::error)
                .orElse(Single.just(publishResult)))
        .doOnSubscribe(ignored -> timer.start())
        .doOnSuccess(ignored -> timer.stopWithSuccess())
        .doOnError(ignored -> timer.stopWithFailure());
```

If you do not need this lazy property (or other reactive properties) of the Single, then it is fine to use the Async API:
```java
timer.start();
asyncClient.publish(message)
        .whenComplete((publishResult, throwable) -> {
            if (throwable != null) {
                timer.stopWithFailure();
            } else {
                timer.stopWithSuccess();
            }
        });
```

As it is possible to solve your problem by applying a few reactive operations, we want to avoid to add extra methods to the client interfaces, so please let us know if it is really important for you. Generally we are up to add convenience methods.

(We experienced a similar problem with latency tracking of publishes, so we are considering to add a context object to Publish messages.)

**Comment by tchupp** [comment]
Thank you for the detailed response!

We indeed were recording latency incorrectly, although our metrics didn't change significantly after correcting the issue (which I believe was due to us not reusing subscriptions, and always subscribing immediately after creating the Single.)

I believe the Context object would be really useful!
I remember when Spring was messing around with a Reactive Streams implementation, they provided the ability to add properties to a "context" object that was carried along with the stream.

```
Single.just(1)
  .context(ctx -> ctx.put("startTime", Instant.now())
  .subscribe((value, ctx) -> {
    println(value); // Prints 1
    println(ctx); // Prints { "startTime": 1234 }
});
```
(might be known as a Writer Monad)

I'm assuming this isn't quite what you mean, since this would involve changing how RxJava works.

What would the interface look like for publishing with a context object?

My first guess would be something along the lines of
```
queue = PublishProcessor.create();

mqttRxClient.publish(queue)
  .doOnNext(wrapper -> wrapper.context.put("stopTime", Instant.now())

...

context = new CustomContext();
context.put("id", UUID.randomUUID());
context.put("startTime", Instant.now());

queue.onNext(new PublishWrapper<CustomContext>(message, context));
```

Also, we discovered (while implementing a stop gap until reconnect is released) that the MqttPublish object that is returned in the MqttPublishResult is not the same object that you pass in the publish flow, and the .equals() and .hashCode() are therefore different!
I see that these issues are already address, but haven't been merged yet.
